// hardhat-finder/dist/src/constants/task.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_RUN_ON_COMPILE = exports.DEFAULT_NO_COMPILE = exports.DEFAULT_COMPACT = exports.DEFAULT_PRETTIFY = exports.DEFAULT_COLORIFY = exports.DEFAULT_INCLUDE_DEPENDENCIES = exports.DEFAULT_MAX_STRING_LENGTH = exports.DEFAULT_DEPTH = exports.DEFAULT_OUTPUTS = exports.SUPPORTED_OUTPUTS = void 0;
// TODO: reorder the outputs
exports.SUPPORTED_OUTPUTS = [
    "artifact",
    "build-info",
    "abi",
    "metadata",
    "bytecode",
    "bytecode-runtime",
    "link-references",
    "link-references-runtime",
    "immutable-references",
    "immutable-references-runtime",
    "source-code",
    "dependencies-source-code",
    "dependencies-info",
    "settings",
    "solc-version",
    "ast",
    "method-identifiers",
    "opcodes",
    "opcodes-runtime",
    "storage-layout",
    "fully-qualified-name",
    "developer-document",
    "user-document",
    "generated-sources",
    "generated-sources-runtime",
    "source-map",
    "source-map-runtime",
];
exports.DEFAULT_OUTPUTS = exports.SUPPORTED_OUTPUTS;
exports.DEFAULT_DEPTH = Infinity;
exports.DEFAULT_MAX_STRING_LENGTH = Infinity;
exports.DEFAULT_INCLUDE_DEPENDENCIES = false;
exports.DEFAULT_COLORIFY = false;
exports.DEFAULT_PRETTIFY = false;
exports.DEFAULT_COMPACT = false;
exports.DEFAULT_NO_COMPILE = false;
exports.DEFAULT_RUN_ON_COMPILE = false;
